### PR TITLE
[CARBONDATA-368]Imporve performance of dataframe loading

### PR DESCRIPTION
**Pull request by QiangCai**
1. support concurrently reading dataframe in CSVInput step

2. fix bug for reading dataframe in kettle thread

3. customize DataLoadPartitionCoalescer to repartition input DataFrame without shuffle.

### CODE REVIEW

**Comment by jackylk** [comment]
Some testcase failed after rebase to master

**Comment by QiangCai** [comment]
I have fix test case error.
now CI PASS
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/595/

File: integration/spark/src/main/scala/org/apache/carbondata/spark/util/CarbonScalaUtil.scala [hunk 1]
```diff
@@ -164,4 +165,53 @@ object CarbonScalaUtil extends Logging {
...
   }
 
+  def getString(value: Any,
+      serializationNullFormat: String,
+      delimiterLevel1: String,
+      delimiterLevel2: String,
+      format: SimpleDateFormat,
+      level: Int = 1): String = {
+    value == null match {
+      case true => serializationNullFormat
```
Comment (jackylk) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/util/CarbonScalaUtil.scala]: please use if else

Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/util/CarbonScalaUtil.scala]: fixed

**Comment by jackylk** [comment]
please rebase

**Comment by QiangCai** [comment]
Rebase done

**Comment by QiangCai** [comment]
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/682/

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataLoadRDD.scala [hunk 2]
```diff
@@ -548,77 +552,53 @@ class DataFrameLoaderRDD[K, V](
   override protected def getPartitions: Array[Partition] = firstParent[Row].partitions
 }
 
+class PartitionIterator(partitionIter: Iterator[DataLoadPartitionWrap[Row]],
+    carbonLoadModel: CarbonLoadModel,
+    context: TaskContext) extends  JavaRddIterator[JavaRddIterator[Array[String]]] {
```
Comment (jackylk) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataLoadRDD.scala]: remove extra space

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 3]
```diff
@@ -932,7 +942,8 @@ object CarbonDataRDDFactory {
           loadDataFile()
         }
         val newStatusMap = scala.collection.mutable.Map.empty[String, String]
-        status.foreach { eachLoadStatus =>
+        if (status.nonEmpty) {
+          status.foreach { eachLoadStatus =>
           val state = newStatusMap.get(eachLoadStatus._1)
```
Comment (jackylk) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: incorrect indentation

File: processing/src/main/java/org/apache/carbondata/processing/csvreaderstep/CsvInput.java [hunk 4]
```diff
@@ -384,21 +383,75 @@ public boolean processRow(StepMetaInterface smi, StepDataInterface sdi) throws K
...
+            }
+          }
+        }
+      } catch (Exception e) {
+        LOGGER.error(e, "Scan rdd during data load is terminated due to error.");
+        throw e;
+      }
+      return null;
+    }
+  };
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/csvreaderstep/CsvInput.java]: remove ;

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataLoadRDD.scala (hunk 2 above)
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataLoadRDD.scala]: fixed

File: processing/src/main/java/org/apache/carbondata/processing/csvreaderstep/CsvInput.java (hunk 4 above)
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/csvreaderstep/CsvInput.java]: fixed

**Comment by jackylk** [comment]
LGTM
CI passed
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/700/

**Pull request closed without merging**